File: medium/javascript/0011-container-with-most-water.js
/**
 * O(n^2)
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  let max = 0;

  for (let i = 0; i < height.length; i++) {
    for (let j = i + 1; j < height.length; j++) {
      if (i === j) {
        continue;
      }

      const area = Math.min(height[i], height[j]) * (j - i);
      if (area > max) {
        max = area;
      }
    }
  }

  return max;
};

/**
 * O(n)
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  let left = 0;
  let right = height.length - 1;
  let max = -Infinity;
  while (left < right) {
    const area = Math.min(height[left], height[right]) * (right - left);
    max = Math.max(max, area);
    if (height[left] > height[right]) {
      right--;
    } else {
      left++;
    }
  }
  return max;
};

File: medium/go/0011-container-with-most-water.go
package main

func maxArea(height []int) int {
    left := 0
	right := len(height)-1
    maxArea := 0

    for left < right {
        area := right - left

        if height[left] < height[right]{
            area *= height[left]
            left++
        }else{
            area *= height[right]
            right--
        }

        if area > maxArea {
            maxArea = area
        }
    } 

    return maxArea
}